train.py

```diff
@@ -321,38 +324,40 @@ def prepare_task(args, devices):
             task.load_dataset(valid_sub_split, combine=True, epoch=0)
 
         # Build models and criteria to print some metadata
-        model_parallel = dp.DataParallel(
-            lambda: task.build_model(args), device_ids=devices)
+        torch.manual_seed(args.seed)
         model, criterion = task.build_model(args), task.build_criterion(args)
-        print_model_criterion(model, criterion, args)
-        del model, criterion
-
-        # Build trainers
-        trainers = {
-            device: Trainer(args, task, model, task.build_criterion(args), xla=True)
-            for device, model in zip(model_parallel.devices, model_parallel.models)
-        }
-        lr = trainers[devices[0]].get_lr()
+        xm.master_print(model)
+        xm.master_print('| model {}, criterion {}'.format(
+            args.arch, criterion.__class__.__name__))
+        xm.master_print('| num. model params: {} (num. trained: {})'.format(
+            sum(p.numel() for p in model.parameters()),
+            sum(p.numel() for p in model.parameters() if p.requires_grad)))
+        model = model.to(xla_device)
+        trainer = Trainer(args, task, model, criterion, xla=True)
+        lr = trainer.get_lr()
 
         # Load the latest checkpoint if one is available and restore the
         # corresponding train iterator
-        extra_state, epoch_itr = checkpoint_utils.load_checkpoint(
-            args, trainers[devices[0]])
-        if extra_state is not None:
-            # checkpoint detected, load saved model weights to all devices
-            xu.eprint(
-                'checkpoint detected, device 0 meters need to be '
-                're-loaded to device'
-            )
-            checkpoint_utils.load_checkpoint_tpu(args, trainers, devices[0])
+        # set distributed args here to shard data
+        trainer.args.distributed_rank = xm.get_ordinal()
+        trainer.args.distributed_world_size = xm.xrt_world_size()
+        extra_state, epoch_itr = checkpoint_utils.load_checkpoint(args, trainer)
+        trainer.args.distributed_rank = 0
+        trainer.args.distributed_world_size = 1
         valid_subsets = args.valid_subset.split(',')
-        return task, trainers, model_parallel, epoch_itr, lr, valid_subsets
+        ordinal = xm.get_ordinal(defval=-1)
+        device_str = (
+            str(xla_device) if ordinal < 0 else
+            '{}/{}'.format(xla_device, ordinal
+        )
+        return task, trainer, model, epoch_itr, lr, valid_subsets, device_str
 
-    def train_loop_fn(model, loader, device, context):
-        trainer = trainers[str(device)]
-        stats, log_output = None, None
-        tracker = xm.RateTracker()
+    def train_loop_fn(device, trainer, loader):
+        stats, log_output, tracker = None, None, xm.RateTracker()
         for i, samples in enumerate(loader):
+            if i+1 == len(loader._loader._loader):
+                #FIXME: last batches are incomplete
+                break
```

Comment:
@dlibenzi, do we have an api to drop last in `parallel_loader` in this case? How do we handle

1. different length loaders
2. last batches (of different size)

Reply:
Making sure the 2nd question doesn't get lost: the last batch loader will produce will be of different length. To avoid recompilations, is there a way to drop that safely?